// revai-node-sdk/src/models/RevAiApiError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidStateError = exports.ResourceNotFoundOrUnsupportedApiError = exports.ForbiddenAccessError = exports.InvalidParameterError = exports.RevAiApiError = void 0;
class RevAiApiError {
    constructor(e) {
        if (e.response) {
            this.statusCode = e.response.status;
            this.details = e.response.data || '';
        }
    }
}
exports.RevAiApiError = RevAiApiError;
class InvalidParameterError extends RevAiApiError {
    constructor(e) {
        super(e);
        this.parameters = e.response.data.parameters;
    }
}
exports.InvalidParameterError = InvalidParameterError;
class ForbiddenAccessError extends RevAiApiError {
    constructor(e) {
        super(e);
        this.parameters = e.response.data.parameters;
    }
}
exports.ForbiddenAccessError = ForbiddenAccessError;
class ResourceNotFoundOrUnsupportedApiError extends RevAiApiError {
    constructor(e) {
        super(e);
        this.parameters = e.response.data.parameters;
    }
}
exports.ResourceNotFoundOrUnsupportedApiError = ResourceNotFoundOrUnsupportedApiError;
class InvalidStateError extends RevAiApiError {
    constructor(e) {
        super(e);
        this.currentValue = e.response.data.current_value;
        this.allowedValues = e.response.data.allowed_values;
    }
}
exports.InvalidStateError = InvalidStateError;
